src/language_features/inlay_hints.rs

```diff
@@ -0,0 +1,105 @@
+    buf_line_count: u32,
+}
+
+pub enum ExperimentalInlayHintRequest {}
+
+// For now, rust analyzer implements this as experimental/inlayHint
+impl Request for ExperimentalInlayHintRequest {
+    type Params = InlayHintParams;
+    type Result = Option<Vec<InlayHint>>;
+    const METHOD: &'static str = "experimental/inlayHints";
+}
+
+pub fn experimental_inlay_hints(meta: EditorMeta, params: EditorParams, ctx: &mut Context) {
+    let params = InlayHintsOptions::deserialize(params).unwrap();
+    let req_params = InlayHintParams {
+        work_done_progress_params: Default::default(),
+        text_document: TextDocumentIdentifier {
+            uri: Url::from_file_path(&meta.buffile).unwrap(),
+        },
+        range: Range::new(Position::new(0, 0), Position::new(params.buf_line_count, 0)),
+    };
+    ctx.call::<ExperimentalInlayHintRequest, _>(meta, req_params, move |ctx, meta, response| {
+        inlay_hints_response(meta, response.unwrap_or_default(), ctx)
+    });
+}
+
+pub fn inlay_hints(meta: EditorMeta, params: EditorParams, ctx: &mut Context) {
+    let params = InlayHintsOptions::deserialize(params).unwrap();
+    let req_params = InlayHintParams {
+        work_done_progress_params: Default::default(),
+        text_document: TextDocumentIdentifier {
+            uri: Url::from_file_path(&meta.buffile).unwrap(),
+        },
+        range: Range::new(Position::new(0, 0), Position::new(params.buf_line_count, 0)),
+    };
+    ctx.call::<InlayHintRequest, _>(meta, req_params, move |ctx, meta, response| {
+        inlay_hints_response(meta, response.unwrap_or_default(), ctx)
+    });
+}
+
+pub fn inlay_hints_response(meta: EditorMeta, inlay_hints: Vec<InlayHint>, ctx: &mut Context) {
+    let document = match ctx.documents.get(&meta.buffile) {
+        Some(document) => document,
+        None => return,
+    };
+    let ranges = inlay_hints
+        .into_iter()
+        .map(
+            |InlayHint {
+                 position,
+                 label,
+                 padding_left,
+                 padding_right,
+                 ..
+             }| {
+                let position =
+                    lsp_position_to_kakoune(&position, &document.text, ctx.offset_encoding);
+                let label = match label {
+                    InlayHintLabel::String(s) => s,
+                    InlayHintLabel::LabelParts(parts) => {
+                        parts.iter().map(|x| x.value.as_str()).collect()
+                    }
+                };
+                let padding_left = if padding_left.unwrap_or(false) {
+                    " "
+                } else {
+                    ""
+                };
+                let padding_right = if padding_right.unwrap_or(false) {
+                    "{Default} "
+                } else {
+                    ""
+                };
+                // Escape markup inside label
+                let label = escape_tuple_element(&label).replace('{', "\\{");
```

Comment:
The doubly escaped backslashes are why I inlined the markup escape - this works, your suggestion does not, since it doubly escapes the backslashes, undoing the markup escaping

Reply:
wow I had no idea. This means that if the label contains \{} then there is no way to render it properly; we are either losing the backslash or the braces.
The markup seems more important, so LGTM

Reply:
I can't reproduce the problem with my suggestion anymore.
OTOH, the current version fails to display inlay hints if I add something like `let label = r#"123\{456"#;` above.

This command illustrates that it works fine to double backslashes at every level of escaping (in addition to escaping { and |)

```
kak -e '
	exec %{%dihello world<esc>}
	declare-option range-specs my_ranges
	add-highlighter global/ replace-ranges my_ranges
	# raw value:                      \{
	#    -> escaped for markup:     \\\{
	#      -> escaped for tuple: \\\\\\{
	set-option global my_ranges %val{timestamp} %<1.5+0|\\\\\\{>
	# renders as hell\{o world
'
```

will commit a fix